bundles/org.openhab.automation.scriptenginefactory.jython/src/main/java/org/openhab/automation/scriptenginefactory/jython/JythonScriptEngineFactory.java

```diff
@@ -0,0 +1,114 @@
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.automation.scriptenginefactory.jython;
+
+import java.io.File;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.TreeSet;
+
+import javax.script.ScriptEngine;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.smarthome.config.core.ConfigConstants;
+import org.openhab.core.automation.module.script.AbstractScriptEngineFactory;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Deactivate;
+
+/**
+ * This is an implementation of {@link ScriptEngineFactory} for Jython.
+ *
+ * @author Scott Rushworth - Initial contribution
+ */
+@Component(service = org.openhab.core.automation.module.script.ScriptEngineFactory.class)
+@NonNullByDefault
+public class JythonScriptEngineFactory extends AbstractScriptEngineFactory {
+
+    private static final String SCRIPT_TYPE = "py";
+    private static javax.script.ScriptEngineManager ENGINE_MANAGER = new javax.script.ScriptEngineManager();
+    private static final String DEFAULT_PYTHON_PATH = new StringBuilder(ConfigConstants.getConfigFolder())
+            .append(File.separator).append("automation").append(File.separator).append("lib").append(File.separator)
+            .append("python").toString();
+
+    @Activate
+    public JythonScriptEngineFactory() {
+        logger.debug("Loading JythonScriptEngineFactory");
+        String home = JythonScriptEngineFactory.class.getProtectionDomain().getCodeSource().getLocation().toString()
+                .replace("file:", "");
+        System.setProperty("python.home", home);
+
+        String existingPythonPath = System.getProperty("python.path");
+        if (existingPythonPath == null || existingPythonPath.isEmpty()) {
+            System.setProperty("python.path", DEFAULT_PYTHON_PATH);
+        } else if (!existingPythonPath.contains(DEFAULT_PYTHON_PATH)) {
+            TreeSet<String> newPythonPathList = new TreeSet<>(
+                    new ArrayList<String>(Arrays.asList(existingPythonPath.split(File.pathSeparator))));
+            newPythonPathList.add(DEFAULT_PYTHON_PATH);
+            String newPythonPath = String.join(File.pathSeparator, newPythonPathList);
+            System.setProperty("python.path", newPythonPath);
+        }
+
+        StringBuilder newPythonCachedir = new StringBuilder(System.getenv("OPENHAB_USERDATA")).append(File.separator)
+                .append("tmp");
+        System.setProperty("python.cachedir", newPythonCachedir.toString());
```

Comment:
I think you can use `ConfigConstants.getUserDataFolder()` instead of reading the environment variable "OPENHAB_USERDATA". And replace the `StringBuilder` by the `Paths.get()` method.

```suggestion
        System.setProperty("python.cachedir", Paths.get(ConfigConstants.getUserDataFolder(), "tmp").toString());
```

Reply:
Much cleaner, thank you! Do you think OPENHAB_USERDATA/tmp/ is the best location for this? For example, I could go down to OPENHAB_USERDATA/tmp/jython/. I think stashing this in the bundle's directory structure would be cleaner, but I didn't come up with a clean way to do it. Do you know if there are other addons that use a directory for caching files?

Reply:
I've moved the python.cachedir to $OPENHAB_USERDATA/cache/org.openhab.automation.scriptenginefactory.jython/cachedir.